### PR TITLE
feat: Spider web scraper & crawler tool

### PR DESCRIPTION
**Pull request by WilliamEspegren**
# Description

Add [Spider](https://spider.cloud/), the [fastest](https://github.com/spider-rs/spider/blob/main/benches/BENCHMARKS.md#benchmark-results) open source scraper & crawler that returns LLM-ready data.

Twitter: [@WilliamEspegren](https://x.com/WilliamEspegren)

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

I have created a new workflow in the UI then tested it manually with all the different parameters.

image

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

This PR was inspired by #3819

### CODE REVIEW

**Comment by WilliamEspegren** [comment]
@bowenliang123 could you check this one out? Let me know if you want any changes or help otherwise :)

**Comment by takatost** [comment]
Hi @WilliamEspegren, I tried out [Spider](https://spider.cloud/), but neither the SIGN IN nor REGISTER buttons are bringing up the login screen as expected. It's happening on both Chrome and Safari.

**Comment by WilliamEspegren** [comment]
> Hi @WilliamEspegren, I tried out [Spider](https://spider.cloud/), but neither the SIGN IN nor REGISTER buttons are bringing up the login screen as expected. It's happening on both Chrome and Safari.

Hi @takatos! Thank you for mentioning this issue. Here is the link instead: https://spider.cloud/login

**Comment by WilliamEspegren** [comment]
@crazywoola @takatost just checking the status on this, if you want any help just ping me :)

File: api/core/tools/provider/builtin/spider/spiderApp.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+class Spider:
+    def __init__(self, api_key: Optional[str] = None):
+        """
+        Initialize the Spider with an API key.
+
+        :param api_key: A string of the API key for Spider. Defaults to the SPIDER_API_KEY environment variable.
+        :raises ValueError: If no API key is provided.
+        """
+        print(api_key)
```
Comment (takatost) [inline api/core/tools/provider/builtin/spider/spiderApp.py]: Please remove the printing of api key

**Review by takatost — CHANGES_REQUESTED** [review]

File: api/core/tools/provider/builtin/spider/spiderApp.py (hunk 1 above)
Comment (WilliamEspegren) [reply api/core/tools/provider/builtin/spider/spiderApp.py]: Thank you for noticing this!

**Review by takatost — APPROVED** [review]
Thanks!!

**Comment by ivuai** [comment]
@WilliamEspegren is this tool still working in latest version? because it doesn't work for me when I enter API key it gives me this error: argument of type 'NoneType' is not iterable

**Comment by WilliamEspegren** [comment]
Hey @ivuai! How does your parameters look like, and what nodes are you using?

**Comment by ivuai** [comment]
I don't use any node, I'm just simply trying to add API key in tools and I don't use any parameters. Here is the screenshot. 
Capture987

**Comment by WilliamEspegren** [comment]
Thank you @ivuai! Weird, I will look and test it out tomorrow :)

**Comment by WilliamEspegren** [comment]
@ivuai the API was somehow down. I have added error checks to display this more clearly if that ever happens again, thanks for mentioning this 👏

**Pull request merged**